WordPress/gutenberg
# BorderControl: Replace style picker with ToggleGroupControl

## What?

Replaces the custom implementation of the border style picker with a standard ToggleGroupControl.

## Why?

Simpler code and improved consistency across similar-looking components (like animations).

## How?

The behavior has been slightly modified, prompted by a limitation in ToggleGroupControl's [value-handling logic](https://github.com/WordPress/gutenberg/blob/84715c01148344c318344978508b5d442526ed0c/packages/components/src/toggle-group-control/toggle-group-control/utils.ts#L21) (related: #47473).

As a backward-compatible way of determining whether it's being used in a controlled or uncontrolled way, on every user-triggered change the ToggleGroupControl compares its new `value` with the previous `value`, and stays in uncontrolled mode if the `value` was unchanged.

BorderControl uses ToggleGroupControl in controlled mode for its border style picker. It has a value sanitization feature (`shouldSanitizeBorder`) that resets the border style to `undefined` if a user tries to pick a border style without setting a border width or color. This means that, even though we need to use ToggleGroupControl in controlled mode, the `value` passed to ToggleGroupControl can start as `undefined` and still be `undefined` after a user-triggered change. ToggleGroupControl interprets this as being in uncontrolled mode, and now the internal value ignores the `value` prop.

However, even if we fixed this limitation so we can replicate the original behavior in trunk, it is actually not a good user interaction because it feels like the buttons are unresponsive.

https://github.com/WordPress/gutenberg/assets/555336/831999d7-3e6c-46bf-b47c-3ac76c78a7dd

When the border styles can't be set, that needs to be communicated to the user in some way, rather than just have the buttons seemingly not respond to clicks. The best way would probably be to implement a `disabled` state for ToggleGroupControl (#57862), and perhaps couple that with `help` text that explains why it's disabled.

For the purposes of this PR however, I propose a more incremental workaround/improvement which is to simply hide the ToggleGroupControl until it's ready to be set.

It's also worth mentioning that the only instance of BorderControl with `shouldSanitizeBorder` enabled in the Gutenberg app is the BorderBoxControl in unlinked mode, so the impact should be low.

## Testing Instructions

1. Go to the Storybook story for BorderControl. When `shouldSanitizeBorder` is true, the border style picker should only be displayed when it can actually be set. 
2. Go to the story for BorderBoxControl. The border style picker should always be visible, as before.

## Screenshots or screencast 

https://github.com/WordPress/gutenberg/assets/555336/2448a8d9-1a3f-4fff-947b-130c1d7c44c8
